tests/0011-produce_batch.c

```diff
@@ -280,8 +281,136 @@ static void test_partitioner (void) {
+    if (!rkt)
+        TEST_FAIL("Failed to create topic: %s\n",
+                  rd_strerror(errno));
+    
+    /* Create messages */
+    rkpartition_counts = calloc(sizeof(int), topic_num_partitions);
+    dr_partition_count = calloc(sizeof(int), topic_num_partitions);
+    rkmessages = calloc(sizeof(*rkmessages), msgcnt);
+    for (i = 0 ; i < msgcnt ; i++) {
+        int *msgidp = malloc(sizeof(*msgidp));
+        *msgidp = i;
+        rd_snprintf(msg, sizeof(msg), "%s:%s test message #%i",
+                    __FILE__, __FUNCTION__, i);
+        
+        rkmessages[i].payload = rd_strdup(msg);
+        rkmessages[i].len     = strlen(msg);
+        rkmessages[i]._private = msgidp;
+        rkmessages[i].partition = rand() % 3;
+        rkpartition_counts[rkmessages[i].partition]++;
+    }
+    
+    r = rd_kafka_produce_batch(rkt, partition, RD_KAFKA_MSG_F_PARTITION | RD_KAFKA_MSG_F_FREE,
+                               rkmessages, msgcnt);
+    
+    /* Scan through messages to check for errors. */
+    for (i = 0 ; i < msgcnt ; i++) {
+        if (rkmessages[i].err) {
+            failcnt++;
+            if (failcnt < 100)
+                TEST_SAY("Message #%i failed: %s\n",
+                         i,
+                         rd_kafka_err2str(rkmessages[i].err));
+        }
+    }
+    
+    /* All messages should've been produced. */
+    if (r < msgcnt) {
+        TEST_SAY("Not all messages were accepted "
+                 "by produce_batch(): %i < %i\n", r, msgcnt);
+        if (msgcnt - r != failcnt)
+            TEST_SAY("Discrepency between failed messages (%i) "
+                     "and return value %i (%i - %i)\n",
+                     failcnt, msgcnt - r, msgcnt, r);
+        TEST_FAIL("%i/%i messages failed\n", msgcnt - r, msgcnt);
+    }
+    
+    free(rkmessages);
+    TEST_SAY("Per-message partition: "
+             "Produced %i messages, waiting for deliveries\n", r);
+    
+    msgcounter = msgcnt;
+    /* Wait for messages to be delivered */
+    test_wait_delivery(rk, &msgcounter);
+    
+    for(i = 0; i < topic_num_partitions; i++) {
+        if (dr_partition_count[i] != rkpartition_counts[i]) {
+            TEST_FAIL("messages were not sent to designated partitions");
+        }
+    }
+    
+    free(rkpartition_counts);
+    free(dr_partition_count);
+    
+    if (msgcounter != 0)
+        TEST_FAIL("Still waiting for %i/%i messages\n",
+                  msgcounter, msgcnt);
+    
+    /* Destroy topic */
+    rd_kafka_topic_destroy(rkt);
+    
+    /* Destroy rdkafka instance */
+    TEST_SAY("Destroying kafka instance %s\n", rd_kafka_name(rk));
+    rd_kafka_destroy(rk);
+    
+    TEST_SAY("removing topic %s", topic_name);
+    test_kafka_topics("--delete --topic %s", topic_name);
+    free(topic_name);
+    
+    return;
+}
+
+
 int main_0011_produce_batch (int argc, char **argv) {
         test_single_partition();
         test_partitioner();
+        test_per_message_partition_flag();
```

Comment:
add two small negative tests that checks that per-message partition is NOT honoured when _F_PARTITION is NOT set:
 * one test with partition == UA, but message..partition==1, make sure at least some messages make it to another partition
 * one test with partition == 1 but message..partition==2, make sure that all messages make it to partition 1. For this just change test_single_partition so that partition=1.
E.g.: pr

Reply:
will implement this in the next commit